pkg/lib/saml/service.go

```diff
@@ -304,7 +308,65 @@ func (s *Service) IssueSuccessResponse(
 
 	response.Assertion = assertion
 
-	// TODO(saml): Sign the response
+	// Sign the assertion
+	// Reference: https://github.com/crewjam/saml/blob/193e551d9a8420216fae88c2b8f4b46696b7bb63/identity_provider.go#L833
+	signingContext, err := s.idpSigningContext()
+	if err != nil {
+		return nil, err
+	}
+
+	assertionEl := response.Assertion.Element()
+
+	signedAssertionEl, err := signingContext.SignEnveloped(assertionEl)
+	if err != nil {
+		return nil, err
+	}
+
+	assertionSigEl := signedAssertionEl.ChildElements()[len(signedAssertionEl.ChildElements())-1]
+	response.Assertion.Signature = assertionSigEl
+
+	// Sign the response
+	responseEl := response.Element()
+	signedResponseEl, err := signingContext.SignEnveloped(responseEl)
+	if err != nil {
+		return nil, err
+	}
+
+	responseSigEl := signedResponseEl.ChildElements()[len(signedResponseEl.ChildElements())-1]
+	response.Signature = responseSigEl
 
 	return response, nil
 }
+
+func (s *Service) idpSigningContext() (*dsig.SigningContext, error) {
+	// Create a cert chain based off of the IDP cert and its intermediates.
+	activeCert, ok := s.SAMLIdpSigningMaterials.FindSigningCert(s.SAMLConfig.Signing.KeyID)
+	if !ok {
+		panic("unexpected: cannot find the corresponding idp key by id")
+	}
+
+	var signingContext *dsig.SigningContext
+	var rsaPrivateKey rsa.PrivateKey
+	err := activeCert.Key.Raw(&rsaPrivateKey)
+	if err != nil {
+		panic(err)
+	}
+
+	keyPair := tls.Certificate{
+		Certificate: [][]byte{activeCert.Certificate.Data()},
+		PrivateKey:  &rsaPrivateKey,
+		Leaf:        activeCert.Certificate.X509Certificate(),
+	}
```

Comment:
Instead of using TLSCertKeyStore, we can borrow MemoryX509KeyStore from https://github.com/russellhaering/goxmldsig/blob/v1.4.0/keystore.go#L31

Using TLS here does not seem appropriate, we are not doing anything related to TLS here.

Reply:
I copied the implementation from https://github.com/crewjam/saml/blob/main/identity_provider.go#L1073-L1078

Maybe let me try to use MemoryX509KeyStore and see if it works.

Reply:
Use NewDefaultSigningContext, and call SetSignatureMethod

Reply:
I found that `MemoryX509KeyStore` cannot be constructed. I will keep using tls keystore here.

Reply:
Or maybe I create a struct which implements x509KeyStore

Reply:
> maybe I create a struct which implements x509KeyStore

Yes. You cannot use `MemoryX509KeyStore` from the library, its fields are all private. But you can copy it very easily. It is just a few lines of code.